xds/internal/client/xds.go

```diff
@@ -140,21 +140,89 @@ func processServerSideListener(lis *v3listenerpb.Listener) (*ListenerUpdate, err
 		return nil, fmt.Errorf("xds: socket_address port does not match the one in name. Got %q, want %q", p, port)
 	}
 
-	// Make sure the listener resource contains a single filter chain. We do not
-	// support multiple filter chains and picking the best match from the list.
-	fcs := lis.GetFilterChains()
-	if n := len(fcs); n != 1 {
-		return nil, fmt.Errorf("xds: filter chains count in LDS response does not match expected. Got %d, want 1", n)
+	var filterChains []*FilterChain
+	for _, fc := range lis.GetFilterChains() {
+		filterChain, err := getFilterChain(fc)
+		if err != nil {
+			return nil, err
+		}
+		filterChains = append(filterChains, filterChain)
+	}
+	defaultFilterChain, err := getFilterChain(lis.GetDefaultFilterChain())
+	if err != nil {
+		return nil, err
+	}
+	if len(filterChains) == 0 && defaultFilterChain == nil {
+		return nil, fmt.Errorf("xds: no supported filter chains and no default filter chain")
+	}
+	return &ListenerUpdate{
+		FilterChains:       filterChains,
+		DefaultFilterChain: defaultFilterChain,
+	}, nil
+}
+
+// getFilterChain parses the filter chain proto and converts it into the local
+// representation. If fc contains unsupported filter chain match fields, a nil
+// FilterChain object and a nil error are returned. If fc does not parse or
+// contains other invalid data, an non-nil error is returned.
+func getFilterChain(fc *v3listenerpb.FilterChain) (*FilterChain, error) {
+	if fc == nil {
+		return nil, nil
+	}
+
+	// If the match criteria contains unsupported fields, skip the filter chain.
+	fcm := fc.GetFilterChainMatch()
+	if fcm.GetDestinationPort().GetValue() != 0 ||
+		fcm.GetServerNames() != nil ||
+		(fcm.GetTransportProtocol() != "" && fcm.TransportProtocol != "raw_buffer") ||
+		fcm.GetApplicationProtocols() != nil {
+		return nil, nil
+	}
+
+	// Extract the supported match criteria.
+	var dstPrefixRanges []net.IP
+	for _, pr := range fcm.GetPrefixRanges() {
+		cidr := fmt.Sprintf("%s/%d", pr.GetAddressPrefix(), pr.GetPrefixLen().GetValue())
+		ip, _, err := net.ParseCIDR(cidr)
+		if err != nil {
+			return nil, fmt.Errorf("xds: failed to parse destination prefix range: %+v", pr)
+		}
+		dstPrefixRanges = append(dstPrefixRanges, ip)
+	}
+	var srcType SourceType
+	switch fcm.GetSourceType().String() {
+	case "SAME_IP_OR_LOOPBACK":
+		srcType = SourceTypeSameOrLoopback
+	case "EXTERNAL":
+		srcType = SourceTypeExternal
```

Comment:
Strings? 👀 

I guess let's have our own consts for these if they aren't defined in the protos somehow (e.g. is there a separate but matching enum we can use?).

Reply:
Sorry, missed this comment.
So, we usually try not to have a dependency on the xds protos in the `XxxUpdate` messages that are sent from the `xdsClient` to upper layers, which is why I created an enum on our side mirroring the enums in the xds protos.

Reply:
> So, we usually try not to have a dependency on the xds protos in the `XxxUpdate` messages that are sent from the `xdsClient` to upper layers

What is the reason for this?  It seems like a waste to redefine types where existing types are suitable.

Reply:
One of the reasons is to make the upper layers agnostic of the actual xds API (transport and resource) versions.